File: kirun-js/src/engine/function/system/date/EpochToTimestamp.ts
import { Schema } from '../../../json/schema/Schema';
import { SchemaType } from '../../../json/schema/type/SchemaType';
import { TypeUtil } from '../../../json/schema/type/TypeUtil';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { MapUtil } from '../../../util/MapUtil';
import { AbstractFunction } from '../../AbstractFunction';
import { AbstractDateFunction } from './AbstractDateFunction';

export class EpochToTimestamp extends AbstractFunction {
    private readonly signature: FunctionSignature;
    private readonly isSeconds: boolean;
    private readonly paramName: string;

    constructor(name: string, isSeconds: boolean) {
        super();
        this.paramName = `epoch${isSeconds ? 'Seconds' : 'Milliseconds'}`;
        this.isSeconds = isSeconds;
        this.signature = new FunctionSignature(name)
            .setNamespace(Namespaces.DATE)
            .setParameters(
                new Map([
                    [
                        this.paramName,
                        Parameter.of(
                            this.paramName,
                            new Schema()
                                .setName(this.paramName)
                                .setType(
                                    TypeUtil.of(
                                        SchemaType.LONG,
                                        SchemaType.INTEGER,
                                        SchemaType.STRING,
                                    ),
                                ),
                        ),
                    ],
                ]),
            )
            .setEvents(
                new Map([
                    [
                        AbstractDateFunction.EVENT_TIMESTAMP_NAME,
                        AbstractDateFunction.EVENT_TIMESTAMP,
                    ],
                ]),
            );
    }

    public getSignature(): FunctionSignature {
        return this.signature;
    }

    protected internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        const epoch = parseInt(context.getArguments()?.get(this.paramName));
        const timestamp = this.isSeconds ? epoch * 1000 : epoch;
        if (isNaN(timestamp)) {
            throw new Error(`Please provide a valid value for ${this.paramName}.`);
        }
        return Promise.resolve(
            new FunctionOutput([
                EventResult.outputOf(
                    MapUtil.of(
                        AbstractDateFunction.EVENT_TIMESTAMP_NAME,
                        new Date(timestamp).toISOString(),
                    ),
                ),
            ]),
        );
    }
}

File: kirun-py/src/kirun_py/function/system/date/epoch_to_timestamp.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import TYPE_CHECKING

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.json.schema.type.schema_type import SchemaType
from kirun_py.json.schema.type.type_util import TypeUtil
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.function_signature import FunctionSignature
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces
from kirun_py.function.system.date.abstract_date_function import AbstractDateFunction

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters


class EpochToTimestamp(AbstractFunction):

    def __init__(self, name: str, is_seconds: bool) -> None:
        super().__init__()
        self._is_seconds = is_seconds
        self._param_name = 'epochSeconds' if is_seconds else 'epochMilliseconds'

        self._signature = (
            FunctionSignature(name)
            .set_namespace(Namespaces.DATE)
            .set_parameters({
                self._param_name: Parameter.of(
                    self._param_name,
                    Schema()
                    .set_name(self._param_name)
                    .set_type(TypeUtil.of(SchemaType.LONG, SchemaType.INTEGER, SchemaType.STRING)),
                ),
            })
            .set_events({
                AbstractDateFunction.EVENT_TIMESTAMP.get_name(): AbstractDateFunction.EVENT_TIMESTAMP,
            })
        )

    def get_signature(self) -> FunctionSignature:
        return self._signature

    async def internal_execute(self, context: FunctionExecutionParameters) -> FunctionOutput:
        args = context.get_arguments() or {}
        raw_epoch = args.get(self._param_name)

        try:
            epoch = int(raw_epoch)
        except (ValueError, TypeError):
            raise ValueError('Please provide a valid value for {}.'.format(self._param_name))

        timestamp_ms = epoch * 1000 if self._is_seconds else epoch
        dt = datetime.fromtimestamp(timestamp_ms / 1000.0, tz=timezone.utc)

        return FunctionOutput([
            EventResult.output_of(
                {AbstractDateFunction.EVENT_TIMESTAMP_NAME: dt.isoformat()}
            ),
        ])
